core/src/main/java/jenkins/model/Nodes.java

```diff
@@ -131,10 +131,11 @@ public void addNode(final @Nonnull Node node) throws IOException {
         if (node != oldNode) {
             // TODO we should not need to lock the queue for adding nodes but until we have a way to update the
             // computer list for just the new node
+            AtomicReference<Node> old = new AtomicReference<>();
```

Comment:
AIUI you are introducing this variable because `oldNode` is obtained without a lock and so might not be the actual node that we end up replacing. If so, then I think on line 152 `oldNode` should be replaced with `old.get()` for the same reason (although that could still be incorrect as mentioned on line 143).

Reply:
> AIUI you are introducing this variable because `oldNode` is obtained without a lock and so might not be the actual node that we end up replacing.

Not as such, I just needed a writable cell to stash the result of `put`, and the `Queue.withLock` overrides take closures rather than the `AutoCloseable` trick so I could not just use a blank final variable.

> on line 152 `oldNode` should be replaced with `old.get()` for the same reason

Perhaps. That would be a different change though. This PR is merely fixing the notification.